### PR TITLE
./scripts/quartus

### PR DESCRIPTION
**Pull request by wallclimber21**
A copy of ./scripts/vivado and then adapted for quartus

Notes:
- I couldn't get the modelsim eval to work, so I fell back to iverilog
- I didn't understand the speed adjustment algorithm in tabtest.sh, so I replaced it with binary search...
- I didn't run 'make synth_system' to completion because it's stalled forever in the mapping phase. Quartus once again doesn't recognize the memory, so it's crunching on 131072 FFs... Even after setting FAST_MEMORY=0, it still doesn't recognize the memory.

To come later: a 'make table.txt' run for various Altera components. Since the current RTL doesn't tickle Quartus into recognizing MLAB RAMs, it's probably better to wait until that's resolved?

### CODE REVIEW

File: scripts/quartus/synth_system.tcl [hunk 1]
```diff
@@ -0,0 +1,17 @@
+
+read_verilog system.v
+read_verilog ../../picorv32.v
+read_xdc synth_system.xdc
+
+synth_design -part xc7a35t-cpg236-1 -top system
```
Comment (cliffordwolf) [inline scripts/quartus/synth_system.tcl]: That can't be right.. ;)

**Comment by cliffordwolf** [comment]
> so I replaced it with binary search...

It was a binary search. You replaced it with a different implementation of a binary search. You can either run bisection by keeping track of a min and max and each iteration is (what you are doing):

    position = (min+max) / 2
    if (condition(position))
        min = position;
    else
        max = position;

Or you can keep track of the center position and a radius (what I am doing):

    if (condition(position))
        position += radius;
    else
        position -= radius;
    radius = radius / 2

They are the same. Invariant properties:

    position = (min+max) / 2
    radius = (max-min) / 2
    min = position - radius
    max = position + radius

File: scripts/quartus/synth_system.tcl (hunk 1 above)
Comment (wallclimber21) [inline scripts/quartus/synth_system.tcl]: Lingering unused old file. Deleted.

**Comment by wallclimber21** [comment]
Hmm. I thought the intention was indeed binary search, but when I ran it, the progression was 50, 58, 60, 62, 61, where timing was failing for each one. There was probably an inverted compare in my code...

**Comment by cliffordwolf** [comment]
> the progression was 50, 58, 60, 62, 61, where timing was failing for each one.

That's the expected behavior for failing timing. The number is the clock period (in 0.1 ns steps), so a higher number is a slower clock. Assuming the timing fails at 3.4 ns (initial min), and assuming it is ok at 6.6ns (initial max), then this is the expected sequence of center positions you'd check in a binary search.

**Comment by cliffordwolf** [comment]
> then this is the expected sequence of center positions you'd check in a binary search.

ok, they are not. ;) It should be 50, 58, 62, 64, 65.

Did you run with init conditions speed=50 and step=8? That's at least what this sequence of numbers suggests.

there is a special hack in `tabtest.sh` that does not half the step size from 8 to 4 when timing fails in the first two iterations (3.0 ns and 3.8 ns). That makes sure the ceiling is high enough for slower devices. It is essentially a performance optimization, reducing the number of iterations by one for most devices without loosing support for the slowest devices in the range.

    [ $speed -eq 38 ] || step=$((step / 2))

Maybe something went wrong with this code in your port? Whatever, it does not really matter..  ;)

**Pull request merged**